File: 377. Combination Sum IV.py
class Solution:
    def combinationSum4(self, nums, target):
    # top down
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        cache = {0 : 1}
        def helper(target):
            if target in cache:
                return cache[target]
            ans = 0
            for num in nums:
                if target - num >= 0:
                    ans += helper(target - num)
            cache[target] = ans
            return ans
        return helper(target)
        
class Solution:
    def combinationSum4(self, nums, target):
    # bottom up
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        ans = [0] * (target + 1)
        ans[0] = 1
        for i in range(1, len(ans)):
            for num in nums:
                if i - num < 0: continue
                ans[i] += ans[i - num]
        return ans[-1]
                        

File: 377. Combination Sum IV.cpp
class Solution {
public:
// top down
    int combinationSum4(vector<int>& nums, int target) {
        unordered_map<int, int> dict{{0, 1}};
        return helper(dict, nums, target);
    }
    int helper(unordered_map<int, int>& dict, vector<int>& nums, int target) {
        if(dict.count(target)) return dict[target];
        int ans = 0;
        for(int n : nums) {
            if(target - n >= 0) ans += helper(dict, nums, target - n);
        }
        dict[target] = ans;
        return ans;
    }
};


class Solution {
public:
// bottom up
    int combinationSum4(vector<int>& nums, int target) {
        vector<int> ans(target + 1, 0);
        ans[0] = 1;
        for(int i = 1; i <= target; ++i) {
            for(int n : nums) {
                if(i - n >= 0) ans[i] += ans[i - n];
            }
        }
        return ans.back();
    }
};
                        